RGB Blood Reworks + Blood generator

Poojawa opened the pull request:
## About The Pull Request

Ho shit folks, I actually figured it all out. Species now have RGB blood effects, due to their blood type. Default and Human is red, of course. Lizards have a more pinkish tone, Slime people have gained hearts and blood - mix with toxin to get Slimejelly! Their blood is a cyan as default, though I have plans to eventually have it inherit their primary color. 

IPCs have a blackish-brown hydraulic fluid, Bugs have a yellowish tone. Xeno people have a green.

A dark blue synthetics blood type has been added, acquired by building a blood bank generator (Sprites courtesy of the always amazing Toriate, I merely added some overlays)

The new machine takes filled blood bags of any type, and converts the blood inside into synthetics at a bonus rate. Synthetics are a one to one conversion, because that should be obvious! Upgrading the machine's micro manipulators will get even more synthetics per unit of blood. 

Footsteps and gibs have been made grey scale, and take the color of their blood type. each species has their own gibtype as well, so no more generic blood-red humans for slime people. Xenos explode with their own sprites also. Eventually I'll unlock the secrets to making custom footprints, but it's good to ship it now.

I had to gut the forensics component, because I wasn't really able to understand how to work with it in the way that I needed it to work. Literally every other codebase uses blood_DNA lists on their objects anyway.

## Why It's Good For The Game

Variety is the spice of life, and just seeing ONLY red blood was kinda silly. It was also silly that slime people could work in Virology with no ill effects. Both IPCs and Slimepeople can now be affected by viruses. They also don't drop dead the moment they're converted to a different species by way of changeling TF stings. 

![](https://i.gyazo.com/76f141a0c7db26aabf899f5a8dfa1ab8.png)

![](https://i.gyazo.com/b4f2a9a39eaa69779e8bf9add0f80491.png)

![](https://i.gyazo.com/1f8b6236f481b1fd55320320c2f440a3.png)

<3

## Changelog
:cl: Poojawa
add: Added RGB blood effects, know whose blood this is by color!
add: Added Synthetics blood, and a machine that produces it. A universal donor for medical to heal people with. Make it via medical protolathe!
fix: With hearts, slimepeople shouldn't die from the effects of not having a heart.
fix: Plasma vessels from xenomorphs will restore blood if you're on resin.
/:cl:

Poojawa committed:
Synthetics blood and greyscale blood effects

Poojawa committed:
compile fixes except we're memes still

why yes, I'm making the bloods all subtypes of each other at this point. I should probably change the way stuff is colored. but, meh. effort.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
pretty sure I'm onto it

Poojawa committed:
greyscale digi shoe blood

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Synthetic and RNG blood

Poojawa committed:
oh god oh fuck oh god it's snowballing

Poojawa committed:
Some slight fixes

Poojawa committed:
system's mostly working

need to polish out the get_blood_dna_list runtime

Poojawa committed:
what to do when footprints are being a bitch.

why, we get rid of them :honk:

Poojawa committed:
I need to do something else for a bit pending reviews

Poojawa committed:
whatever Travis, it isn't even enabled, fuck off

Poojawa committed:
makes travis shutup

Poojawa committed:
I really dunno at this point, weh

Poojawa committed:
go away Travis

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
adjust and reset things

Poojawa committed:
well we have footprints now, xenoblood back to blood subtype.

Poojawa committed:
progress, just need to fiddle with footprints again

Poojawa committed:
get blood id is my foe now.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
okay we're starting to pass our blood type to bloodmeme now

STILL JUST FUCKING FOOTSTEPS LEFT THO GOD FUCKING DAMNIT

Poojawa committed:
Fixes Xenomorph blood related travis errors

Poojawa committed:
Ok this time all of them are nitpicked

Poojawa committed:
duplicate definitions? really? wtf

Poojawa committed:
it's the one part that I keep saying I just need to get around to

Poojawa committed:
Update human_movement.dm

Poojawa committed:
u h g. I hate this

Poojawa committed:
I think I got it.

Also gibs should be better too.

Poojawa committed:
as it turns out, Creatures don't have minds

Poojawa committed:
I don't even fucking know anymore

Poojawa committed:
just need the stupid component to work properly now

Poojawa committed:
port tg 42691 for component transfer enforcement

Poojawa committed:
map editing is nicer with notepad ++ at least

Poojawa committed:
do we even use this shitty away mission

Poojawa committed:
,

Poojawa committed:
Update wildwest.dmm

Poojawa committed:
Blood colors happens to footprints now.

Different prints not working. splatter effect not working, I'm just sick of looking at this right now.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
reverts forensics components. If I can't figure it out, MURDER IT

Poojawa committed:
oh right.

Poojawa committed:
aaaaAAAAAAAA

NOW WHY DO I HAVE INFINITE FOOTPRINTS AAAAAAAAAAAAA

Poojawa committed:
eh, was going somewhere with this

IPC gibspawner, looking into infinite footprints

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
lel

Poojawa committed:
fuckit, I'm just gunna wholesale port from baycode I guess

Poojawa committed:
these too please thanks

Poojawa committed:
Revert "these too please thanks"

This reverts commit bc80c036791d16bc33d3d841d212003b31c7e102.

Poojawa committed:
Revert "fuckit, I'm just gunna wholesale port from baycode I guess"

This reverts commit 11d10d38052d7cd2077a5b698f4c555b01780354.

Poojawa committed:
Revert "lel"

This reverts commit 9e01a078ec5f8d96e1cb7d9b2c882f1e46ce03a1.

Poojawa committed:
KISS principal

Poojawa committed:
just need to fine tweak gibs and we're gucci

Poojawa committed:
Gibs spawn correctly. Just need to figure out the overlays now

Poojawa committed:
god I can't believe I'm matrixing gib overlays. save me PLEASE

Poojawa committed:
NYEH. fuckin' memes

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into syntheticbloods

Poojawa committed:
ok

Poojawa committed:
holy FUCK I got gib overlays working. Why wasn't it before?

No one will ever know.

Poojawa committed:
neyh

Poojawa committed:
some good work here

Poojawa committed:
itsworking.avi

Poojawa committed:
bug blood, ipc/slime hearts, blood gen protolathe

we're all gucchi now boys

Poojawa commented on _maps/RandomZLevels/moonoutpost19.dmm:
```diff
@@ -1970,33 +1970,21 @@
-	dir = 8;
-	icon_state = "ltrails_1"
-	},
+/obj/effect/decal/cleanable/blood/footprints/tracks/wheels,
```
also I'm pretty sure most of these got fucked by my 'just make it fucking compile' but I don't think people will really care that much tbh

Poojawa committed:
I accidentally all the maps. whoops

Poojawa committed:
Delete footprints_and_rag.dm

just bloat from a previous time

deathride58 reviewed (changes requested):
DNM until forensics component is readded. If you have trouble with this, then point me in the direction of a commit or branch made prior to your removal of the forensics component.

deathride58 commented on code/game/atoms.dm:
```diff
@@ -243,23 +248,22 @@
-	. = "[icon2html(src, user)] [thats? "That's ":""][get_examine_name(user)]"
-
 /atom/proc/examine(mob/user)
-	to_chat(user, get_examine_string(user, TRUE))
```
Also tiny nitpick here but what's the reason for this?

Poojawa replied:
it was part of the examine replacement text.

Which was part of the component.

Ghommie replied:
Regardless, those two functions above helped reduced the verbosity of many examination lines, such as a mob's worn items, and I suggest they should be preserved if possible.

deathride58 replied:
Unresolving since get_examine_string is a much cleaner way of handling this, and you didn't really bother with trying to replace any of the other uses of get_examine_string, meaning all of those other use cases are missing the ability to display blood-stained status

Ghommie commented on code/__DEFINES/citadel_defines.dm:
```diff
@@ -110,6 +110,7 @@
 #define isaquatic(A) (is_species(A, /datum/species/aquatic))
 #define isinsect(A) (is_species(A, /datum/species/insect))
 #define isxenoperson(A) (is_species(A, /datum/species/xeno))
+#define isstartjelly(A) (is_species(A, /datum/species/jelly/roundstartslime))
```
Could you move these is[x] helpers to is_helpers.dm, pretty please?

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Ghommie commented on code/datums/dna.dm:
```diff
@@ -303,6 +303,14 @@
 	if(!dna.species)
 		var/rando_race = pick(GLOB.roundstart_races)
 		dna.species = new rando_race()
+		if(isjellyperson(dna.species))
```
Wouldn't it be more efficient to simply set the respective specieses' exotic_bloodtype var to these new snowflakes and then replace all of this with 
```
		if(dna.species.exotic_bloodtype)
			dna.blood_type = dna.species.exotic_bloodtype`
```
Edit: Just follow up the second half since you seemengly have already done the first.

Ghommie commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -270,7 +269,16 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 
 	if(exotic_bloodtype && C.dna.blood_type != exotic_bloodtype)
-		C.dna.blood_type = exotic_bloodtype
+		if(isjellyperson(C))
```
If you followed up the previous comment, this shouldn't be necessary.

Ghommie commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -270,7 +269,16 @@ GLOBAL_LIST_EMPTY(roundstart_races)
+		if(isxenoperson(C))
+			C.dna.blood_type = "X*"
+		if(islizard(C))
+			C.dna.blood_type = "L"
```
Also datum/species/lizard/var/exotic_bloodtype has the exact same value already.
This and the line above are useless either way.

Ghommie commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -311,7 +319,16 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 /datum/species/proc/on_species_loss(mob/living/carbon/human/C, datum/species/new_species, pref_load)
 	if(C.dna.species.exotic_bloodtype)
-		C.dna.blood_type = random_blood_type()
+		if(isjellyperson(C))
```
Bad copypasta, the blood type should revert to a random normal blood type upon species loss.

Ghommie commented on code/game/atoms.dm:
```diff
@@ -243,23 +248,22 @@
 	if(desc)
 		to_chat(user, desc)
+	// *****RM
+	//to_chat(user, "[name]: Dn:[density] dir:[dir] cont:[contents] icon:[icon] is:[icon_state] loc:[loc]")
```
Isn't this one hacky debugging helper to_chat() that probably should be removed?

Ghommie commented on code/game/atoms.dm:
```diff
@@ -344,18 +349,125 @@
+	return transfer_blood_dna(blood_dna)
+
+/obj/item/add_blood(list/blood_dna)
+	if(!..())
```
Do this instead:
```
	. = ..()
	if(!.)
```
This way you can also pass down the parent return value to possible subtypes' own definitions of this function. Little coding standard.

deathride58 replied:
Unresolving since this is a much cleaner method of handling input from parent procs
EDIT: since github doesnt display the original comment in the review, here's a copy-paste

> Do this instead:
> 
> ```
> 	. = ..()
> 	if(!.)
> ```
> This way you can also pass down the parent return value to possible subtypes' own definitions of this function. Little coding standard.

Ghommie commented on code/game/atoms.dm:
```diff
@@ -344,18 +349,125 @@
+
+/obj/item/add_blood(list/blood_dna)
+	if(!..())
+		return FALSE
```
if you followed the previous comment then just remove `FALSE` since it's already implied to be false.

Ghommie commented on code/game/atoms.dm:
```diff
@@ -344,18 +349,125 @@
+	if(!..())
+		return FALSE
+	add_blood_overlay()
+	return TRUE //we applied blood to the item
```
if you followed the suggestion, just remove this line altogether, it return `.` by default at the very end.

Poojawa committed:
likely fixes the examine bug lol

Ghommie commented on code/game/machinery/bloodbankgen.dm:
```diff
@@ -0,0 +1,335 @@
+		if(bag.reagents.total_volume)
+			var/mutable_appearance/filling_overlay = mutable_appearance(icon, "input-reagent")
+
+			var/percent = round((bag.reagents.total_volume / bag.volume) * 100)
```
Just saying, these sort of switches could be avoided with `filling_overlay.icon_state = "input-reagent[FLOOR(bag.reagents.total_volume / bag.volume, 10)]`.

Except that requires a consistent size for each "level", which is not the case here.

deathride58 replied:
Unresolving since this is worth consideration. Replacing this with a oneline `icon_state = "iconstate[FLOOR(round(cap/value), 10)"` would do wonders to help improve the code quality here

EDIT: copypaste of original comment
> Just saying, these sort of switches could be avoided with `filling_overlay.icon_state = "input-reagent[FLOOR(bag.reagents.total_volume / bag.volume, 10)]`.
> 
> Except that requires a consistent size for each "level", which is not the case here.

Ghommie commented on code/game/objects/effects/decals/cleanable.dm:
```diff
@@ -69,10 +70,9 @@
 	if(ishuman(O))
 		var/mob/living/carbon/human/H = O
-		if(H.shoes && blood_state && bloodiness && !HAS_TRAIT(H, TRAIT_LIGHT_STEP))
+		if(H.shoes && blood_state && bloodiness && (!HAS_TRAIT(H,TRAIT_LIGHT_STEP) || !H.mind.assigned_role == "Detective"))
```
In the long run, the traits system fullfils these sort of checks better than these mind.assigned_roles, now that we have the mind_traits var for jobs.
So add a new trait, call it "TRAIT_CRIME_INVESTIGATOR", add it to line 152 of `job_types/security.dm` and then do:
```
		if(H.shoes && blood_state && bloodiness && (!HAS_TRAIT(H,TRAIT_LIGHT_STEP|) || (!H.mind || !HAS_TRAIT(H.mind, TRAIT_CRIME_INVESTIGATOR)))
```
Otherwise just add an !H.mind safety check, because this is likely to cause null.assigned_role runtimes.

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+	if(istype(L) && has_gravity(loc))
+		if(ishuman(L))
+			var/mob/living/carbon/human/H = L
+			if(H.mind.assigned_role == "Detective") //Gumshoe perks yo
```
Trait over mind.assigned_role as I said before.

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+
+/obj/effect/decal/cleanable/blood/gibs/old/Initialize(mapload, list/datum/disease/diseases)
+	. = ..()
+	setDir(pick(1,2,4,8))
```
```suggestion
	setDir(pick(GLOB.cardinals))
```

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+	var/body_colors = "#e3ba84"	//a default color just in case.
+
+/obj/effect/decal/cleanable/blood/gibs/Initialize(mapload, list/datum/disease/diseases)
+	. = ..()
```
I'm seeing fairly similar steps being done on every gibs subtypes Initialize(). I find it pretty stupid when something like this could be done instead:
```
	if(gibs_reagent_id)
		reagents.add_reagent(gibs_reagent_id, 5)
	if(gibs_bloodtype && !length(blood_DNA))
		add_blood(list("Non-human DNA" = gibs_bloodtype))
```

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+	reagents.add_reagent("liquidgibs", 5)
+	if(!blood_DNA.len)
+		add_blood(list("Non-human DNA" = "A+"))
+	update_icon()
```
Remove this proc definition if you are going to follow up the above suggestion.
But also remember to set up the two new vars for this subtype.

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+	reagents.add_reagent("liquidgibs", 5)
+	if(!blood_DNA.len)
+		add_blood(list("Non-human DNA" = "L"))
+	update_icon()
```
Idem.

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+	reagents.add_reagent("liquidslimegibs", 5)
+	if(!blood_DNA.len)
+		add_blood(list("Non-human DNA" = "GEL"))
+	update_icon()
```
Idem.

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+
+/obj/effect/decal/cleanable/blood/gibs/synth/Initialize(mapload, list/datum/disease/diseases)
+	. = ..()
+	reagents.add_reagent("liquidsyntheticgibs", 5)
```
Idem.

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+	reagents.add_reagent("liquidoilgibs", 5)
+	if(!blood_DNA.len)
+		add_blood(list("Non-human DNA" = "HF"))
+	update_icon()
```
Idem.

Ghommie commented on code/game/objects/effects/decals/cleanable/aliens.dm:
```diff
@@ -1,71 +1,72 @@
+/obj/effect/decal/cleanable/blood/gibs/xeno/Initialize(mapload, list/datum/disease/diseases)
+	. = ..()
+	reagents.add_reagent("liquidxenogibs", 5)
+	update_icon()
```
Same.

Ghommie commented on code/game/objects/effects/decals/cleanable/gibs.dm:
```diff
@@ -0,0 +1,240 @@
+	. = ..()
+	setDir(pick(1,2,4,8))
+	icon_state += "-old"
+	add_blood(list("Non-human DNA" = "A+"))
```
Also remove this line in favour of the var, should you be following the suggestion.

Ghommie commented on code/__HELPERS/_cit_helpers.dm:
```diff
@@ -93,6 +93,72 @@ GLOBAL_LIST_INIT(dildo_colors, list(//mostly neon colors
 		"Purple"	= "#e300ff"//purple
 		))
 
+GLOBAL_LIST_INIT(noodle_taurs, list(
```
move all these global lists to `_globalvars/lists/mobs.dm`

Ghommie commented on code/modules/mob/living/blood.dm:
```diff
@@ -138,17 +138,27 @@
+							C.reagents.add_reagent("bonehurtingjuice", amount * 0.5)
+					else
+						C.blood_volume = min(C.blood_volume + round(amount, 0.1), BLOOD_VOLUME_MAXIMUM)
+						return 1
```
```suggestion
					return 1
```
stopping the target's blood volume from being resotred from incompatible bloods.

Also, this else if list is sucky, but I'm getting tired of reviewing this.

Ghommie commented on code/modules/mob/living/blood.dm:
```diff
@@ -217,12 +229,18 @@
+	if(ishuman(C))
+		var/mob/living/carbon/human/H = C
+		if(H.dna.species.exotic_blood)
+			return H.dna.species.exotic_blood
```
Remove these useless new lines, dna is a carbon var.

Ghommie commented on code/modules/mob/living/blood.dm:
```diff
@@ -217,12 +229,18 @@
+		var/mob/living/carbon/human/H = C
+		if(H.dna.species.exotic_blood)
+			return H.dna.species.exotic_blood
 	if(dna.species.exotic_blood)
```
```suggestion
	if(dna?.species?.exotic_blood)
```
Not all carbons (such aliens) have dna.

Ghommie commented on code/modules/mob/living/brain/brain.dm:
```diff
@@ -21,6 +24,14 @@
 	if(!stored_dna.species)
 		var/rando_race = pick(GLOB.roundstart_races)
 		stored_dna.species = new rando_race()
+		if(isjellyperson(stored_dna.species))
```
See the comment about species exotic blood types. Replace all of this with:
```
		if(stored_dna.species.exotic_bloodtype)
			stored_dna.blood_type = stored_dna.species.exotic_bloodtype
```

Ghommie commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -486,3 +486,22 @@ It's fairly easy to fix if dealing with single letters but not so much with comp
+	. = BLOOD_COLOR_HUMAN
+	switch(type)
+		if("U")//Universal blood; a bit orange
+			. = "#db3300"
```
Make this a define.

Ghommie commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -486,3 +486,22 @@ It's fairly easy to fix if dealing with single letters but not so much with comp
+		if("GEL")// slimepeople blood, rgb 0, 255, 144
+			. = BLOOD_COLOR_SLIME
+		if("BUG")
+			. = "#a37c0f"
```
Same here.

Ghommie reviewed (commented):
A little tired right now, reviewed 3/4 of the thing more or less.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Impliments Ghom review, reverts blood list back to component

Poojawa committed:
reverts the add_blood back to add_blood_DNA

Poojawa committed:
I'm about to go fucking APE

None of this is actually viable, Footprints aren't taking proper color, aren't updating their color. It's just default red. screm

Poojawa committed:
meh.

Poojawa committed:
I'm just waiting for a better idea before dumping this hot garbage again

Poojawa committed:
Forensics re-destroyed, but better this time.

All lists are LAZYINITLIST, cleaning code is preserved, examine is preserved. We should be back in buisness, but this is just a 'holy fuck finally' commit

Poojawa committed:
just need items to actually have their overlays, ree

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
since it conflicted there, let's just fix it here

Poojawa committed:
Update human.dm

Poojawa committed:
Everything's working again, for now.

Poojawa committed:
sure I guess

Poojawa committed:
fixes the IV dripper memes, better formatting too

Poojawa committed:
REEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE

Poojawa committed:
fixes tgui I guess lol

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
blood dna is fixed, your ERP code is too

Poojawa committed:
reverts the majority of clean act because it doesn't work properly otherwise

Poojawa committed:
It's working again, I guess. wtf

Poojawa committed:
Virology's broken as fuck, damnifino why

Poojawa committed:
Fixes missing DNA info for slimes, resolves virology issues

Poojawa committed:
Allows xeno weeds to regenerate lost blood

Honestly this should have been a thing since forever ago but whatever.

Poojawa committed:
Fixes some errant issues with overlays

Poojawa committed:
Revert/replaces a lot of #8024 with bleed_rate

Wearing armor reduces the blood loss you incur, no longer are you instantly deleted on blood.  Dripping blood WILL kill you if you don't solve that issue, medical gauze has been tripled as a QoL for medical staff as a result.

Poojawa committed:
WHAT IN TARNATION

Poojawa committed:
on second thought, let's not try to half ass fancy ideas.

Poojawa committed:
buffs toxinlover blood deletion penalty

Poojawa committed:
Blood Generator QOL, put in hands, select slot to add bloodbag to

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Maps in Blood generators

Poojawa committed:
MIGHT HELP IF SYNTHETICS EXISTED, WHOOPS

Poojawa committed:
Medical assistance treats bleed_rate to stop phantom bleeding

Poojawa committed:
let's be sure people don't have negative bleeding, because that would be weird as fuck.

Poojawa committed:
Ohwait let's not return on a normal process. :^)

Poojawa committed:
hard caps bleeding rate to 20

Poojawa committed:
Cult heals do bleed stopping

blood cap raised because honestly there's so much bleed_rate heal available now too

Poojawa committed:
Human magic revival does bleeding heal

Poojawa committed:
Medibeam heals bleed_rate and radiation

Poojawa committed:
Gauze fully stops bleed rate on application

Poojawa committed:
reverts the bleed changes till they can be better tuned

Poojawa committed:
Update moonoutpost19.dmm

Poojawa committed:
Double checking my work

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
doing the little things

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge branch 'syntheticbloods' of https://github.com/Poojawa/Citadel-Station-13-portworks into syntheticbloods

deathride58 commented on code/game/machinery/washing_machine.dm:
```diff
@@ -59,15 +55,16 @@
 /obj/machinery/washing_machine/proc/wash_cycle()
 	for(var/X in contents)
 		var/atom/movable/AM = X
-		SEND_SIGNAL(AM, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
Readd the clean_act signal, blood isn't the only thing that uses it

Poojawa replied:
Yes it is, actually. the only thing this signal did was clean_blood but with more obfuscation. I absolutely am 140% promising you that I tried every godamn way of preserving it but without the blood component, it didn't work.

deathride58 replied:
What is thermite?

deathride58 commented on code/game/objects/items/twohanded.dm:
```diff
@@ -346,7 +346,8 @@
 		icon_state = "dualsaber[item_color][wielded]"
 	else
 		icon_state = "dualsaber0"
-	SEND_SIGNAL(src, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
The clean_act signal needs to be readded, blood isn't the only thing that uses it

deathride58 commented on code/game/turfs/open.dm:
```diff
@@ -250,7 +250,7 @@
 	for(var/mob/living/simple_animal/slime/M in src)
 		M.apply_water()
 
-	SEND_SIGNAL(src, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_WEAK)
```
Ditto

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -317,7 +317,7 @@
 
 
 /obj/machinery/shower/proc/wash_obj(obj/O)
-	. = SEND_SIGNAL(O, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_WEAK)
```
Ditto

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -328,15 +328,14 @@
 /obj/machinery/shower/proc/wash_turf()
 	if(isturf(loc))
 		var/turf/tile = loc
-		SEND_SIGNAL(tile, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_WEAK)
```
Ditto

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -328,15 +328,14 @@
 
 
 /obj/machinery/shower/proc/wash_mob(mob/living/L)
-	SEND_SIGNAL(L, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_WEAK)
```
Ditto

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -381,7 +380,7 @@
 			else if(H.w_uniform && wash_obj(H.w_uniform))
 				H.update_inv_w_uniform()
 			if(washgloves)
-				SEND_SIGNAL(H, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
Ditto. There's plenty of other instances where you removed the clean_act signal. Readd 'em. You don't have to replace clean_blood() to do it.

deathride58 commented on code/modules/mob/living/blood.dm:
```diff
@@ -141,29 +152,29 @@
 				if(!(blood_data["blood_type"] in get_safe_blood(C.dna.blood_type)))
-					C.reagents.add_reagent("toxin", amount * 0.5)
-					return 1
+					C.reagents.add_reagent("bonehurtingjuice", amount * 0.5)
```
What's the reason for this change? Wouldn't it be better to add a different variant of toxin that deals toxinloss that ignores the toxinlover flag instead of changing this to a meme chem that makes people spam "oof", "ouch", and "owie"?

Poojawa replied:
Slimepeople could end up using any blood for free, massive heals. this was a better solution that still held a viable penalty for not paying attention to your blood type.

Linzolle replied:
is there anything that's stopping you from just... using adjusttoxloss straight up without a reagent inbetween

deathride58 reviewed (changes requested):
Quite a few reviews you marked as resolved were not actually resolved. Let's see if github properly handles adding review comments to resolved reviews. (EDIT: gdi, it doesnt)
Additionally, you need to readd the clean component signal, as blood absolutely is not the only thing that relies on it.

Poojawa committed:
I still fucking hate components.

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
fixes oddities

deathride58 commented on modular_citadel/code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1,11 +1,3 @@
-/datum/reagent/space_cleaner/reaction_obj(obj/O, reac_volume)
-	if(istype(O, /obj/effect/decal/cleanable) || istype(O, /obj/item/projectile/bullet/reusable/foam_dart) || istype(O, /obj/item/ammo_casing/caseless/foam_dart))
```
This bit hasn't been properly demodularized. Space cleaner is supposed to clean up foam darts.

deathride58 commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -27,22 +27,62 @@
+		if((blood_id == "blood" || blood_id == "jellyblood") && (method == INJECT || (method == INGEST && C.dna && C.dna.species && (DRINKSBLOOD in C.dna.species.species_traits))))
 			if(!data || !(data["blood_type"] in get_safe_blood(C.dna.blood_type)))
-				C.reagents.add_reagent("toxin", reac_volume * 0.5)
+				C.reagents.add_reagent("bonehurtingjuice", reac_volume * 0.5)
```
Okay but seriously, bonehurtingjuice does nothing but deal staminaloss and make the person spam "oof", "ouch", and "owie". Dealing toxinloss directly would be very much preferable to this.

Poojawa replied:
I would prefer a different chem than free healing for slimepeople

deathride58 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -583,7 +583,7 @@ All effects don't start immediately, but rather get worse over time; the rate is
 /datum/reagent/consumable/ethanol/bloody_mary/on_mob_life(mob/living/carbon/C)
 	if(C.blood_volume < (BLOOD_VOLUME_NORMAL*C.blood_ratio))
-		C.blood_volume = min((BLOOD_VOLUME_NORMAL*C.blood_ratio), C.blood_volume + 3) //Bloody Mary quickly restores blood loss.
+		C.blood_volume += 2 //Bloody Mary quickly restores blood loss.
```
What's the reason for this?

deathride58 commented on code/modules/mob/living/blood.dm:
```diff
@@ -141,29 +152,29 @@
 				if(!(blood_data["blood_type"] in get_safe_blood(C.dna.blood_type)))
-					C.reagents.add_reagent("toxin", amount * 0.5)
-					return 1
+					C.reagents.add_reagent("bonehurtingjuice", amount * 0.5)
```
Linzolle's proposed change of adjusting the toxinloss directly would be very much preferable to changing this to bonehurtingjuice. Staminaloss and forcefully spamming "oof", "ouch", and "owie" is not a viable penalty.

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -498,7 +500,7 @@
 			H.regenerate_icons()
 		user.drowsyness = max(user.drowsyness - rand(2,3), 0) //Washing your face wakes you up if you're falling asleep
 	else
-		SEND_SIGNAL(user, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
Comsig

Poojawa replied:
handled higher up on the proc

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -398,9 +400,9 @@
-			SEND_SIGNAL(M, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
+			M.clean_blood()
 	else
-		SEND_SIGNAL(L, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
Comsig

Poojawa replied:
This one's handled higher up

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -398,9 +400,9 @@
 		else
 			if(M.wear_mask && wash_obj(M.wear_mask))
 				M.update_inv_wear_mask(0)
-			SEND_SIGNAL(M, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
Comsig

Poojawa replied:
handled higher up on the proc

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -381,7 +383,7 @@
 			else if(H.w_uniform && wash_obj(H.w_uniform))
 				H.update_inv_w_uniform()
 			if(washgloves)
-				SEND_SIGNAL(H, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
Comsig

Poojawa replied:
handled higher up on the proc

deathride58 commented on code/game/objects/items/twohanded.dm:
```diff
@@ -346,7 +346,8 @@
 		icon_state = "dualsaber[item_color][wielded]"
 	else
 		icon_state = "dualsaber0"
-	SEND_SIGNAL(src, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
Comsig

Poojawa replied:
This doesn't need to be handled. It's purely there only as an overlay cleanup.

deathride58 commented on modular_citadel/code/game/objects/items/melee/eutactic_blades.dm:
```diff
@@ -304,7 +304,7 @@
 		var/mob/M = loc
 		M.update_inv_hands()
 
-	SEND_SIGNAL(src, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)//blood overlays get weird otherwise, because the sprite changes. (retained from original desword because I have no idea what this is)
```
Comsig

Poojawa replied:
ditto with that dsword. it's only for blood overlay fuckery

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
does some of the television's demands

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Update other_reagents.dm

Poojawa committed:
ok

Poojawa committed:
why am I making it  ignore injections????

The world may never know

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
readds slime heart

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Helping edgecase things

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
re-readd slime hearts

again

Poojawa committed:
blood sorta works again, needs better fixing tho

Poojawa committed:
Fixes blood transfusion,  QOL to IV drippers

Poojawa committed:
clears a merge conflict

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
Merge remote-tracking branch 'Citadel-Station-13/master' into syntheticbloods

Poojawa committed:
re-adds blood crate

Thalpy commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -947,12 +1069,12 @@
+	overdose_threshold = 30
 	color = "#c2391d"
 
 /datum/reagent/iron/on_mob_life(mob/living/carbon/C)
```
Surely an efficacy reduction by 98x on top of an OD is harsh? Taking 30u would only yield 0.75u total after 2.5 minutes of metabolism. In this case sugar is a better blood treatment. What options are there for people to treat blood loss when medbay ceases to be an option?

Poojawa replied:
Cargo can order synthetic blood bags, and they can be administered with the CMO hypo in single, large doses if needed.

Ghommie replied:
*laughs in bloody mary*

Poojawa committed:
Merge remote-tracking branch 'refs/remotes/Citadel-Station-13/master' into syntheticbloods

deathride58 commented on code/game/atoms.dm:
```diff
@@ -254,14 +260,27 @@
 	. = "[icon2html(src, user)] [thats? "That's ":""][get_examine_name(user)]"
 
 /atom/proc/examine(mob/user)
-	to_chat(user, get_examine_string(user, TRUE))
```
Leaving another comment as part of this review. The functionality of displaying blood as seen in this proc should be rolled into get_examine_string() and this proc rolled back to relying on get_examine_string(), as get_examine_string() is used multiple times elsewhere and has not been properly updated to reflect blood being inherent to the base atom type.

Poojawa replied:
It's literally only used for blood overlay name updating. There's nothing else that uses it currently at all.

deathride58 replied:
![image](https://user-images.githubusercontent.com/6356337/67255322-72600580-f44f-11e9-8ed5-7030e0e65694.png)

Poojawa replied:
And what does get_examine_string do, perchance?

deathride58 replied:
get_examine_string() generates the name for an item, allowing things like blood to show up when you're examining a spaceman without the copypasta associated with the explicit checks for blood. this also allows the examine string to be overridden by individual objects, though this behavior isn't yet used in tgcode. prior to the proc being implemented, every single one of those get_examine_string() calls had their own individual checks for blood, which was extremely messy.

Poojawa replied:
it's still checking each item for the component before the reversion. The current code updates its status the same way the component basically did, just with less obfuscation that the component brings. 

It still isn't used for anything else, and either doesn't apply at all how it was (which was the case quite frequently) or EVERYTHING was blood stained

deathride58 replied:
It doesn't check for an explicit component. It checks for all components, which is how components are designed to work; to have behavior be easily expandable without directly touching the relevant code. The proc just needs to be modified to take into account blood no longer being a component, that's all I'm asking.

Poojawa replied:
It already has, though? it works as intended right now.

deathride58 commented on code/game/atoms.dm:
```diff
@@ -349,18 +370,117 @@
+	return transfer_blood_dna(blood_dna, diseases)
+
+/obj/item/add_blood_DNA(list/blood_dna, list/datum/disease/diseases)
+	if(!..())
```
> Unresolving since this is a much cleaner method of handling input from parent procs
> EDIT: since github doesnt display the original comment in the review, here's a copy-paste
> 
> > Do this instead:
> > ```
> > 	. = ..()
> > 	if(!.)
> > ```
> > 
> > 
> > This way you can also pass down the parent return value to possible subtypes' own definitions of this function. Little coding standard.

Still needs addressing.

Poojawa replied:
it was attempted, and found to not be working.

deathride58 replied:
How exactly does it not work?

Poojawa replied:
> the suggestion of
> 
> . = ..()
> 	if(!.)
> in atom.dm also did not work. it would return the TRUE as a FALSE and then fail the entire proc. because byond.

deathride58 replied:
Was it runtiming? Did you follow Ghommie's other two comments in this proc?

Poojawa replied:
It wasn't run-timing, because it encountered the if false return, so it didn't do anything. Wasn't adding overlays as a result.

The other two comments required it to work at all. I even added debug text to see what the fuck was going on, and it would say shit like 'Atom level returned true' 'item level returned false' I vocalized a great majority of my frustration in the discord.

deathride58 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -381,7 +383,7 @@
 			else if(H.w_uniform && wash_obj(H.w_uniform))
 				H.update_inv_w_uniform()
 			if(washgloves)
-				SEND_SIGNAL(H, COMSIG_COMPONENT_CLEAN_ACT, CLEAN_STRENGTH_BLOOD)
```
The comsig on H still needs readding. Same thing with the two below this. The comsig higher up only cleans the tile.

deathride58 reviewed (changes requested):
Changes I requested a month ago still have not been addressed.

Poojawa committed:
Merge branch 'master' into syntheticbloods

Poojawa committed:
Adds the blood types as actual blood reagents too

Poojawa committed:
Fuck components

Poojawa committed:
if this doesn't work, I won't be fixing it

deathride58 reviewed (approved):
Parts of my review have gone completely unaddressed but yknow what? Fuck it. I'll fix it myself in a followup PR. This has waited long enough.
